File: app/src/main/java/wangdaye/com/geometricweather/main/dialogs/LocationHelpDialog.java
package wangdaye.com.geometricweather.main.dialogs;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.utils.helpers.IntentHelper;
import wangdaye.com.geometricweather.main.MainActivity;

public class LocationHelpDialog {

    public static void show(Activity activity) {
        View view = LayoutInflater
                .from(activity)
                .inflate(R.layout.dialog_location_help, null, false);
        initWidget(
                activity,
                view,
                new MaterialAlertDialogBuilder(activity)
                        .setTitle(R.string.feedback_location_help_title)
                        .setView(view)
                        .show()
        );
    }

    @SuppressLint("SetTextI18n")
    private static void initWidget(Activity activity, View view, AlertDialog dialog) {
        view.findViewById(R.id.dialog_location_help_permissionContainer)
                .setOnClickListener(v -> IntentHelper.startApplicationDetailsActivity(activity));

        view.findViewById(R.id.dialog_location_help_locationContainer)
                .setOnClickListener(v -> IntentHelper.startLocationSettingsActivity(activity));

        view.findViewById(R.id.dialog_location_help_providerContainer)
                .setOnClickListener(v -> IntentHelper.startSelectProviderActivity(activity));

        view.findViewById(R.id.dialog_location_help_manageContainer).setOnClickListener(v -> {
            if (activity instanceof MainActivity) {
                ((MainActivity) activity).setManagementFragmentVisibility(true);
            } else {
                IntentHelper.startMainActivityForManagement(activity);
            }

            dialog.dismiss();
        });
        ((TextView) view.findViewById(R.id.dialog_location_help_manageTitle)).setText(
                activity.getString(R.string.feedback_add_location_manually).replace(
                        "$", activity.getString(R.string.current_location)
                )
        );
    }
}

File: app/src/main/java/org/breezyweather/ui/main/dialogs/LocationHelpDialog.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.ui.main.dialogs

import android.app.Activity
import android.view.LayoutInflater
import androidx.annotation.DrawableRes
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.AlertDialogDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.ListItem
import androidx.compose.material3.ListItemDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import org.breezyweather.R
import org.breezyweather.common.utils.helpers.IntentHelper
import org.breezyweather.ui.main.MainActivity
import org.breezyweather.ui.main.utils.MainThemeColorProvider
import org.breezyweather.ui.theme.compose.BreezyWeatherTheme

object LocationHelpDialog {
    fun show(
        activity: Activity,
    ) {
        val view = LayoutInflater
            .from(activity)
            .inflate(R.layout.dialog_location_help, activity.findViewById(android.R.id.content), true)

        val composeView = view.findViewById<ComposeView>(R.id.location_help_dialog)
        val dialogOpenState = mutableStateOf(true)
        val isDaylight = if (activity is MainActivity &&
            (!activity.isManagementFragmentVisible || activity.isDrawerLayoutVisible)
        ) { // only use the location-based daylight setting when the home fragment is visible
            activity.isDaylight
        } else {
            null
        }

        composeView.setContent {
            BreezyWeatherTheme(
                !MainThemeColorProvider.isLightTheme(activity, daylight = isDaylight)
            ) {
                if (dialogOpenState.value) {
                    AlertDialog(
                        onDismissRequest = {
                            dialogOpenState.value = false
                        },
                        confirmButton = {
                            TextButton(
                                onClick = {
                                    dialogOpenState.value = false
                                }
                            ) {
                                Text(stringResource(R.string.action_close))
                            }
                        },
                        title = {
                            Text(
                                stringResource(R.string.location_message_failed_to_locate)
                            )
                        },
                        text = {
                            Column(
                                modifier = Modifier.fillMaxWidth()
                            ) {
                                DialogListItem(
                                    content = stringResource(
                                        R.string.location_dialog_failed_to_locate_action_select_source
                                    ),
                                    iconId = R.drawable.ic_factory,
                                    onClick = {
                                        IntentHelper.startLocationProviderSettingsActivity(activity)
                                        dialogOpenState.value = false
                                    }
                                )
                                DialogListItem(
                                    content = stringResource(
                                        R.string.location_dialog_failed_to_locate_action_add_manually,
                                        stringResource(R.string.location_current)
                                    ),
                                    iconId = R.drawable.ic_list,
                                    onClick = {
                                        if (activity is MainActivity) {
                                            activity.setManagementFragmentVisibility(true)
                                        } else {
                                            IntentHelper.startMainActivityForManagement(activity)
                                        }
                                        dialogOpenState.value = false
                                    }
                                )
                            }
                        }
                    )
                }
            }
        }
    }

    @Composable
    private fun DialogListItem(
        content: String,
        @DrawableRes iconId: Int? = null,
        onClick: () -> Unit,
    ) {
        ListItem(
            colors = ListItemDefaults.colors(containerColor = AlertDialogDefaults.containerColor),
            modifier = Modifier.clickable { onClick() },
            headlineContent = {
                Text(
                    content,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    style = MaterialTheme.typography.bodyMedium
                )
            },
            leadingContent = iconId?.let {
                {
                    Icon(
                        painterResource(it),
                        contentDescription = null
                    )
                }
            }
        )
    }
}
